Xamarin.Forms.Core/Picker.cs

```diff
@@ -7,246 +7,249 @@
-		void RemoveItems(NotifyCollectionChangedEventArgs e)
-		{
-			int index = e.OldStartingIndex < Items.Count ? e.OldStartingIndex : Items.Count;
-			// TODO: How do we determine the order of which the items were removed
-			foreach (object _ in e.OldItems)
-			{
-				Items.RemoveAt(index--);
-			}
-		}
-
-		void UpdateSelectedIndex(object selectedItem)
-		{
-			string displayMember = GetDisplayMember(selectedItem);
-			int index = Items.IndexOf(displayMember);
-			// TODO Should we prevent call to FindObject since the object is already known
-			// by setting a flag, or otherwise indicate, that we, internally, forced a SelectedIndex changed
-			SelectedIndex = index;
-		}
-
-		void UpdateSelectedItem()
-		{
-			if (SelectedIndex < 0 || SelectedIndex > Items.Count - 1)
-			{
-				SelectedItem = null;
-			}
-			else
-			{
-				SelectedItem = ItemsSource?[SelectedIndex];
-			}
-		}
-	}
+    [RenderWith(typeof(_PickerRenderer))]
+    public class Picker : View
+    {
+        public static readonly BindableProperty TextColorProperty = BindableProperty.Create(nameof(TextColor), typeof(Color), typeof(Picker), Color.Default);
+
+        public static readonly BindableProperty TitleProperty = BindableProperty.Create(nameof(Title), typeof(string), typeof(Picker), default(string));
+
+        public static readonly BindableProperty SelectedIndexProperty = BindableProperty.Create(nameof(SelectedIndex), typeof(int), typeof(Picker), -1, BindingMode.TwoWay,
+            propertyChanged: OnSelectedIndexChanged,
+            coerceValue: CoerceSelectedIndex);
+
+        public static readonly BindableProperty SelectedValueMemberPathProperty = BindableProperty.Create(nameof(SelectedValueMemberPath), typeof(string), typeof(Picker));
+
+        public static readonly BindableProperty ItemsSourceProperty = BindableProperty.Create(nameof(ItemsSource), typeof(IList), typeof(Picker), default(IList), propertyChanged: OnItemsSourceChanged);
+
+        public static readonly BindableProperty DisplayMemberPathProperty = BindableProperty.Create(nameof(DisplayMemberPath), typeof(string), typeof(Picker), propertyChanged: OnDisplayMemberPathChanged);
+
+        public static readonly BindableProperty SelectedItemProperty = BindableProperty.Create(nameof(SelectedItem), typeof(object), typeof(Picker), null, BindingMode.TwoWay, propertyChanged: OnSelectedItemChanged);
+
+        public static readonly BindableProperty DisplayFuncProperty =
+            BindableProperty.Create(
+                nameof(DisplayFunc),
+                typeof(Func<object, string>),
+                typeof(Picker));
+        public Picker()
+        {
+            ((ObservableList<string>)Items).CollectionChanged += OnItemsCollectionChanged;
+        }
+
+        public string DisplayMemberPath
+        {
+            get { return (string)GetValue(DisplayMemberPathProperty); }
+            set { SetValue(DisplayMemberPathProperty, value); }
+        }
+
+        public Func<object, string> DisplayFunc
```

Comment:
Best of both world sounds cool!

On Mon, 22 Aug 2016 at 05:28 Joakim Carselind <notifications@github.com>
wrote:

> In Xamarin.Forms.Core/Picker.cs
> <https://github.com/xamarin/Xamarin.Forms/pull/296#discussion_r75635498>:
>
> > +            BindableProperty.Create(
> > +                nameof(DisplayFunc),
> > +                typeof(Func<object, string>),
> > +                typeof(Picker));
> > +        public Picker()
> > +        {
> > +            ((ObservableList<string>)Items).CollectionChanged += OnItemsCollectionChanged;
> > +        }
> > +
> > +        public string DisplayMemberPath
> > +        {
> > +            get { return (string)GetValue(DisplayMemberPathProperty); }
> > +            set { SetValue(DisplayMemberPathProperty, value); }
> > +        }
> > +
> > +        public Func<object, string> DisplayFunc
>
> Perhaps that would be better. If the converter doesn't return string then
> we could throw an exception. Perhaps include both options?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/xamarin/Xamarin.Forms/pull/296/files/5c1d5e149dc21c58cebf7cdbc6677d1ccec04ed4..49c7876bda4185c699f5fd9b3a66763efca9623c#r75635498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGr-6ldQoJMJpzCqqaknb1mXEdQM8RHsks5qiV29gaJpZM4JiGCM>
> .
>